Dockerfile

```diff
@@ -1,3 +1,37 @@
-FROM starefossen/github-pages:onbuild
+# Get Jekyll build env
+FROM docs/docker.github.io:docs-builder AS builder
 
-CMD jekyll serve -d /_site --watch -H 0.0.0.0 -P 4000
+# Set to the version for this archive
+# You need to set this one more time below too
+ENV VER=v1.4
+
+# Build the docs from this branch
+COPY . /source
+RUN jekyll build --source /source --destination /site/${VER}
+
+# Do post-processing on archive
+COPY --from=docs/docker.github.io:docs-config /scripts/* /usr/bin/
+RUN /usr/bin/fix-archives.sh /site/ ${VER}
+
+# Make an index.html which will redirect / to /${VER}/
+RUN VER=${VER} echo "<html><head><title>Redirect for $VER</title><meta http-equiv=\"refresh\" content=\"0;url='/$VER/'\" /></head><body><p>If you are not redirected automatically, click <a href=\"/$VER/\">here</a>.</p></body></html>" > /site/index.html
+
+# Reset to bare Nginx image
+FROM nginx:alpine
+
+# Set to the version for this archive
+ENV VER=v1.4
+
+# Copy the Nginx config from the docs-config image
+COPY --from=docs/docker.github.io:docs-config /conf/nginx-overrides.conf /etc/nginx/conf.d/default.conf
+
+# Clean out any existing HTML files
+RUN ls -a /usr/share/nginx/html/* && rm -rf /usr/share/nginx/html/*
+
+# Copy the HTML from the builder stage to the default location for Nginx
+COPY --from=builder /site/${VER}/ /usr/share/nginx/html/${VER}/
+COPY --from=builder /site/index.html /usr/share/nginx/html/index.html
+COPY --from=builder /site/index.html /usr/share/nginx/html/404.html
```

Comment:
It's cleaner if the 404.html is generated in the builder image as well, then the "clean" and "copy" steps can simply be:

```Dockerfile
RUN rm -rf /usr/share/nginx/html/*
COPY --from=builder /site /usr/share/nginx/html/
```

Doing so:

- It's no longer needed to have a `$VER` environment variable here (although it may be useful to have it around - also see my `ARG` example above if you don't want to repeat it in each step)
- This build stage no longer has to be aware of what files to copy: separation of concerns; the build generates all the HTML we need; the `nginx:alpine` build-stage only copies the results to the final image.

Reply:
With my comments, the Dockerfile would look like;

```Dockerfile
# Set to the version for this archive
ARG VER=v1.4

# Get Jekyll build env
FROM docs/docker.github.io:docs-builder AS builder

# Make the version accessible to this build-stage
ARG VER

# Build the docs from this branch
COPY . /source
RUN jekyll build --source /source --destination /site/${VER}

# Do post-processing on archive
COPY --from=docs/docker.github.io:docs-config /scripts/* /usr/bin/
RUN /usr/bin/fix-archives.sh /site/ ${VER}

# Make an index.html and 404.html which will redirect / to /${VER}/
RUN echo "<html><head><title>Redirect for ${VER}</title><meta http-equiv=\"refresh\" content=\"0;url='/${VER}/'\" /></head><body><p>If you are not redirected automatically, click <a href=\"/${VER}/\">here</a>.</p></body></html>" > /site/index.html
RUN echo "<html><head><title>Redirect for ${VER}</title><meta http-equiv=\"refresh\" content=\"0;url='/${VER}/'\" /></head><body><p>If you are not redirected automatically, click <a href=\"/${VER}/\">here</a>.</p></body></html>" > /site/404.html


# Reset to bare Nginx image
FROM nginx:alpine

# Make the version accessible to this build-stage, and copy it to an ENV so that it persists in the final image
ARG VER
ENV VER=$VER

# Clean out any existing HTML files, and copy the HTML from the builder stage to the default location for Nginx
RUN rm -rf /usr/share/nginx/html/*
COPY --from=builder /site /usr/share/nginx/html

# Copy the Nginx config from the docs-config image
COPY --from=docs/docker.github.io:docs-config /conf/nginx-overrides.conf /etc/nginx/conf.d/default.conf

# Start Nginx to serve the archive at / (which will redirect to the version-specific dir)
CMD echo -e "Docker docs are viewable at:\nhttp://0.0.0.0:4000"; exec nginx -g 'daemon off;'
```